providers/base/units/led/jobs.pxu

```diff
@@ -347,4 +347,50 @@ estimated_duration: 10
 imports: from com.canonical.plainbox import manifest
 requires: manifest.has_led_indicator == 'True'
 command:
-    led_test.sh -t sysfs -n {path}
+    led_control_test.py -n {path} -d 10
+
+id: led-indicator/gpio-controller-leds
+plugin: resource
+_summary: Gather a list of LED indicators for the device that controls via GPIO chip.
+_description:
+    A LED GPIO devices mapping resource that relies on the user specifying in config variable.
+    Refer to the /sys/kernel/debug/gpio for the gpiochip number
+    Usage of parameter: GPIP_CONTROLLER_LEDS={name1}:{controller1}:{port1} {name2}:{controller1}:{port2} ...
+    e.g. GPIP_CONTROLLER_LEDS=dl14:3:1 dl15:3:2 dl16:3:3
+estimated_duration: 3
+environ: GPIP_CONTROLLER_LEDS
+command:
+    awk '{
+        split($0, record, " ")
+        for (i in record) {
+            split(record[i], data, ":")
+            printf "name: %s\nchip_number: %s\nport: %s\n\n", data[1], data[2], data[3]
+        }
+    }' <<< "$GPIP_CONTROLLER_LEDS"
```

Comment:
This looks a little convoluted; it's a bit hard to read and cannot be tested easily.

How about a tiny Python script that takes `GPIP_CONTROLLER_LEDS` as an argument and does:

```python
def controller_leds_resource(text):
    output = ""
    resource_text = "name: {}\nchip_number: {}\nport: {}\n\n"
    for led in text.split():
        name, chip_number, port = led.split(":")
        output += resource_text.format(name, chip_number, port)
    return output
```

This way you can write a unit test for it.

Reply:
Thanks for the feedback, I have updated the gpio_control_test.py scripts and led resource job. Please review the latest changes.